sdk/core/Azure.Core/src/DynamicData/DynamicData.AllowList.cs

```diff
@@ -0,0 +1,277 @@
+            {
+                if (value == null)
+                {
+                    return true;
+                }
+
+                if (!type.IsGenericType)
+                {
+                    return false;
+                }
+
+                if (type.GetGenericTypeDefinition() != typeof(Dictionary<,>))
+                {
+                    return false;
+                }
+
+                Type[] types = type.GetGenericArguments();
+                if (types[0] != typeof(string))
+                {
+                    return false;
+                }
+
+                if (IsAllowedNonInheritableType(types[1]))
+                {
+                    return true;
+                }
+
+                return IsAllowedEnumerableValue(types[1], ((IDictionary)value).Values);
+            }
+
+            private static bool IsAllowedEnumerableValue(Type elementType, IEnumerable enumerable)
+            {
+                foreach (var item in enumerable)
+                {
+                    if (item == null)
+                    {
+                        continue;
+                    }
+
+                    // Don't allow heterogenous collections
+                    if (item.GetType() != elementType)
+                    {
+                        return false;
+                    }
+
+                    if (!IsAllowedValue(item))
+                    {
+                        return false;
+                    }
+                }
+
+                return true;
+            }
+
+            private static bool IsAllowedAnonymousValue<T>(Type type, T value, Type[]? visited, int depth)
+            {
+                if (!IsAnonymousType(type))
+                {
+                    return false;
+                }
+
+                foreach (PropertyInfo property in type.GetProperties())
+                {
+                    if (IsAllowedLeafType(property.PropertyType))
+                    {
+                        continue;
+                    }
+
+                    object? propertyValue = property.GetValue(value);
+                    if (IsAllowedCollectionValue(property.PropertyType, propertyValue))
+                    {
+                        continue;
+                    }
+
+                    // Detect cycles: trust but verify
+                    if (ContainsType(visited, depth, property.PropertyType))
+                    {
+                        continue;
+                    }
+
+                    // Recurse
+                    visited = AddType(visited, depth, type);
+                    if (!IsAllowedAnonymousValue(property.PropertyType, propertyValue, visited, depth + 1))
+                    {
+                        return false;
+                    }
+
+                    if (depth == 0 && visited != null)
+                    {
+                        // We're back at the top of the stack
+                        ArrayPool<Type>.Shared.Return(visited);
```

Comment:
Are you sure visited won't be used after you return it to the pool? I worry that this routine is called in the loop and the caller might still hold a reference to the array. In general, I cannot tell if this code is correct or not, and so maybe we should just use List<Type> in V1 and optimize it later to a pooled array.

Reply:
Per discussion, we removed the cycle detection, as we were convinced you can't make an assignment that has a cycle in the object graph, given the restriction to anonymous types instead of POCOs and collections that can't contain subtypes of the generic argument type.  This enabled a number of further simplifications.